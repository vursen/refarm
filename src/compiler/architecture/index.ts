export class Context {
  classes: Map<string, Klass> = new Map()
  components: Map<string, Component> = new Map()

  addKlass () {

  }

  addComponent () {

  }
}

export class Klass {

  methods: Map<string, Method>

  trackedProperties: Map<string, TrackedProperty>

  visit () {

  }

  visitMethod () {

  }

  visitTrackedProperty () {

  }
}

export class TrackedProperty {
  // TODO: Type? Klass? Primitive?

  klass: Klass
  context: Context

  reference: Klass

  visit () {

  }
}

export class Method {
  klass: Klass
  context: Context

  calledMethods: Set<Method> = new Set()
  readedTrackedProperties: Set<TrackedProperty> = new Set()
  writedTrackedProperties: Set<TrackedProperty> = new Set()

  visit () {

  }
}

export class Component {
  class: ComponentKlass

  template: ComponentTemplate
}

export class ComponentTemplate {

}

export class ComponentKlass extends Klass {
  methods: Map<string, ComponentMethod>
  attributes: Map<string, ComponentAttribute>

  visit () {
    super.visit()
  }

  visitAttribute () {

  }
}

export class ComponentMethod extends Method {
  klass: ComponentKlass

  readedAttributes: Set<ComponentAttribute> = new Set()

  visit () {
    super.visit()
  }
}

export class ComponentAttribute {
  klass: ComponentKlass
  context: Context

  visit () {

  }
}
