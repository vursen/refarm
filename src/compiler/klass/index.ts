import * as tsMorph from 'ts-morph'

import { Context } from '../context'

import { Method, isMethodDeclaration } from './method'
import { TrackedProperty, isTrackedPropertyDeclaration } from './tracked-property'

export class Klass {
  methods: Map<Method['name'], Method>

  trackedProperties: Map<TrackedProperty['name'], TrackedProperty>

  constructor (
    public node: tsMorph.ClassDeclaration,
    public context: Context
  ) {}

  get name () {
    return this.node.getName()
  }

  visit () {
    for (const member of this.node.getInstanceMembers()) {
      if (isTrackedPropertyDeclaration(member)) {
        this.visitTrackedPropertyDeclaration(member)
      }

      if (isMethodDeclaration(member)) {
        this.visitMethodDeclaration(member)
      }
    }
  }

  visitDeeply () {
    for (const trackedProperty of this.trackedProperties.values()) {
      trackedProperty.visit()
    }

    for (const method of this.methods.values()) {
      method.visit()
    }
  }

  visitMethodDeclaration (node: tsMorph.MethodDeclaration) {
    const method = new Method(
      node,
      this
    )

    this.methods.set(method.name, method)
  }

  visitTrackedPropertyDeclaration (node: tsMorph.PropertyDeclaration) {
    const trackedProperty = new TrackedProperty(
      node,
      this
    )

    this.trackedProperties.set(trackedProperty.name, trackedProperty)
  }

  // TODO: Remove
  // visitInjectDeclaration (node: tsMorph.PropertyDeclaration) {
  //   const typeDeclaration = node.getType()?.getSymbol()?.getDeclarations()?.[0]

  //   if (typeDeclaration && isStoreClassDeclaration(typeDeclaration)) {
  //     const name = node.getName()
  //     const path = typeDeclaration.getSourceFile().getFilePath()

  //     this.injectedStores.set(
  //       name,
  //       this.context.addStoreAtPath(path)
  //     )
  //   }
  // }
}
