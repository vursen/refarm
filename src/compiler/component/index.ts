import * as tsMorph from 'ts-morph'

import { Context } from '../context'

import { Klass } from './klass'
import { Template } from './template'

export class Component {
  klass = new Klass(
    this.scriptSourceFile.getClass(this.name)!,
    this.context
  )

  template = new Template(
    this.templateSourceFile,
    this.context,
    this
  )

  constructor (
    public scriptSourceFile: tsMorph.SourceFile,
    public templateSourceFile: tsMorph.SourceFile,
    public context: Context
  ) {}

  get name () {
    return this.scriptSourceFile.getBaseNameWithoutExtension()
  }

  visit () {
    this.klass.visit()
    this.klass.visitDeeply()

    // this.template.visit()
  }

  // addImportedComponent (path: string) {
  //   const resolvedFilePath = this.context.resolveModuleName(
  //     path,
  //     this.templateSourceFile.getFilePath()
  //   )

  //   if (!resolvedFilePath) {
  //     return
  //   }

  //   const component = this.context.addComponentAtPath(resolvedFilePath)

  //   this.importedComponents.set(
  //     component.name,
  //     component
  //   )
  // }
}
