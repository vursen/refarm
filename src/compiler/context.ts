import * as tsMorph from 'ts-morph'

import { Klass } from './klass'
import { Component } from './component'

export class Context {
  project: tsMorph.Project

  klasses: Map<Klass['node'], Klass> = new Map()

  components: Map<string/* path */, Component> = new Map()

  constructor (options: { tsConfigFilePath: string }) {
    const { tsConfigFilePath } = options

    this.project = new tsMorph.Project({
      tsConfigFilePath,
      addFilesFromTsConfig: false
    })
  }

  resolveModuleName (moduleName: string, containingFile: string) {
    const { resolvedModule } = tsMorph.ts.resolveModuleName(
      moduleName,
      containingFile,
      this.project.getCompilerOptions(),
      this.project.getModuleResolutionHost()
    )

    return resolvedModule?.resolvedFileName
  }

  addKlass (node: Klass['node']) {
    if (this.klasses.has(node)) {
      return this.klasses.get(node)!
    }

    const klass = new Klass(
      node,
      this
    )

    this.klasses.set(node, klass)

    klass.visit()
    klass.visitDeeply()

    return klass
  }

  addComponent (filePath: string) {
    // TODO: Refactor
    const filePathWithoutExtension = filePath.replace(/(\.html|\.ts)$/, '')

    if (this.components.has(filePathWithoutExtension)) {
      return this.components.get(filePathWithoutExtension)!
    }

    const component = new Component(
      this.project.addSourceFileAtPath(`${filePathWithoutExtension}.ts`),
      this.project.addSourceFileAtPath(`${filePathWithoutExtension}.html`),
      this
    )

    this.components.set(filePathWithoutExtension, component)

    component.visit()

    return component
  }
}
