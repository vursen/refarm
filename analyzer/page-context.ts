import * as path from 'path'
import * as tsMorph from 'ts-morph'
import * as svelte from 'svelte/compiler'

import { Store } from './store'
import { Component } from './component'

export class PageContext {
  project: tsMorph.Project

  stores: Map<string, Store> = new Map()

  components: Map<string, Component> = new Map()

  constructor (_pagePath: string) {
    this.project = new tsMorph.Project({
      tsConfigFilePath: path.join(__dirname, '../tsconfig.json'),
      addFilesFromTsConfig: false
    })
  }

  resolveModuleName (moduleName: string, containingFile: string) {
    const { resolvedModule } = tsMorph.ts.resolveModuleName(
      moduleName.replace(/\.ts|\.html/, ''),
      containingFile,
      this.project.getCompilerOptions(),
      this.project.getModuleResolutionHost()
    )

    // console.log(moduleName, containingFile)

    return resolvedModule
  }

  addStoreAtPath (filePath: string) {
    if (this.stores.has(filePath)) {
      return this.stores.get(filePath)
    }

    const store = new Store(
      this.project.addSourceFileAtPath(filePath),
      this
    )

    this.stores.set(filePath, store)

    store.visit()

    return store
  }

  addComponentAtPath (filePath: string) {
    // TODO: Refactor
    const filePathWithoutExtension = filePath.replace(/(\.html|\.ts)$/, '')

    if (this.components.has(filePathWithoutExtension)) {
      return this.components.get(filePathWithoutExtension)
    }

    const component = new Component(
      this.project.addSourceFileAtPath(`${filePathWithoutExtension}.ts`),
      this.project.addSourceFileAtPath(`${filePathWithoutExtension}.html`),
      this
    )

    this.components.set(filePathWithoutExtension, component)

    component.visit()

    return component
  }
}
