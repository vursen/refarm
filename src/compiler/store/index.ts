import * as tsMorph from 'ts-morph'

import { PageContext } from '../page-context'

import {
  isStoreClassDeclaration,
  isStoreStateDeclaration,
  isStoreInjectDeclaration
} from './utils'

import { StateDefinition } from './state-definition'
import { MethodDefinition } from './method-definition'

export class Store {
  injectedStores: Map<string, Store> = new Map()

  stateDefinitions: Map<string, StateDefinition> = new Map()

  methodDefinitions: Map<string, MethodDefinition> = new Map()

  constructor (
    public sourceFile: tsMorph.SourceFile,
    public pageContext: PageContext
  ) {}

  get name () {
    return this.sourceFile.getBaseNameWithoutExtension()
  }

  /**
   * The visitor
   */
  visit () {
    const classDeclaration = this.sourceFile.getClass(this.name)

    // 1. Visit state property declarations
    classDeclaration!.getInstanceMembers().forEach((member) => {
      if (isStoreStateDeclaration(member)) {
        this.visitStateDeclaration(member)
      }
    })

    // 2. Visit method declarations
    classDeclaration!.getInstanceMethods().forEach((member) => {
      this.visitMethodDeclaration(member)
    })

    // 3. Visit inject property declarations
    classDeclaration!.getInstanceMembers().forEach((member) => {
      if (isStoreInjectDeclaration(member)) {
        this.visitInjectDeclaration(member)
      }
    })

    // 4. Visit method definitions deeply
    ;[...this.methodDefinitions.values()].forEach((definition) => {
      definition.visit()
    })
  }

  /**
   * The visitor of inject property declarations
   *
   * ```
   * @Inject someDependency: someType
   * ```
   */
  visitInjectDeclaration (node: tsMorph.PropertyDeclaration) {
    const typeDeclaration = node.getType()?.getSymbol()?.getDeclarations()?.[0]

    if (typeDeclaration && isStoreClassDeclaration(typeDeclaration)) {
      const name = node.getName()
      const path = typeDeclaration.getSourceFile().getFilePath()

      this.injectedStores.set(
        name,
        this.pageContext.addStoreAtPath(path)
      )
    }
  }

  /**
   * The visitor of state property declarations
   *
   * ```
   * @State someProperty = []
   * ```
   */
  visitStateDeclaration (node: tsMorph.PropertyDeclaration) {
    const name = node.getName()

    this.stateDefinitions.set(
      name,
      new StateDefinition(node, this)
    )
  }

  /**
   * The visitor of method declarations
   *
   * ```
   * someMethod (...) { ... }
   * ```
   */
  visitMethodDeclaration (node: tsMorph.MethodDeclaration) {
    const name = node.getName()

    this.methodDefinitions.set(
      name,
      new MethodDefinition(node, this)
    )
  }
}
