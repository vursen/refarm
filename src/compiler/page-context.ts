import * as tsMorph from 'ts-morph'

import { Store } from './store'
import { Component } from './component'

export class PageContext {
  project: tsMorph.Project

  stores: Map<string, Store> = new Map()

  components: Map<string, Component> = new Map()

  constructor (options: { tsConfigFilePath: string }) {
    const { tsConfigFilePath } = options

    this.project = new tsMorph.Project({
      tsConfigFilePath,
      addFilesFromTsConfig: false
    })
  }

  resolveModuleName (moduleName: string, containingFile: string) {
    const { resolvedModule } = tsMorph.ts.resolveModuleName(
      moduleName,
      containingFile,
      this.project.getCompilerOptions(),
      this.project.getModuleResolutionHost()
    )

    return resolvedModule?.resolvedFileName
  }

  addStoreAtPath (filePath: string) {
    if (this.stores.has(filePath)) {
      return this.stores.get(filePath)!
    }

    const store = new Store(
      this.project.addSourceFileAtPath(filePath),
      this
    )

    this.stores.set(filePath, store)

    store.visit()

    return store
  }

  // addTemplateAtPath (filePath: string) {

  // }

  addComponentAtPath (filePath: string) {
    // TODO: Refactor
    const filePathWithoutExtension = filePath.replace(/(\.html|\.ts)$/, '')

    if (this.components.has(filePathWithoutExtension)) {
      return this.components.get(filePathWithoutExtension)!
    }

    const component = new Component(
      this.project.addSourceFileAtPath(`${filePathWithoutExtension}.ts`),
      this.project.addSourceFileAtPath(`${filePathWithoutExtension}.html`),
      this
    )

    this.components.set(filePathWithoutExtension, component)

    component.visit()

    return component
  }
}
