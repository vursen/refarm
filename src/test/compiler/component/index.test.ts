import * as path from 'path'

import { PageContext } from '../../../compiler/page-context'

describe('store', () => {
  const pageContext = new PageContext({
    tsConfigFilePath: path.join(__dirname, '../../../../project/tsconfig.json')
  })

  const component = pageContext.addComponentAtPath(
    path.join(__dirname, '../../../../project/pages/ProductPage.ts')
  )

  const {
    propertyDefinitions,
    importedComponents,
    injectedStores
  } = component

  describe('template', () => {
    it('should parse ast', () => {
      expect(component.template.dump()).toMatchSnapshot()
    })
  })

  describe('visitor', () => {
    // it('should collect imported components', () => {
    //   expect(importedComponents.size).toBe(2)
    //   expect(importedComponents.has('Photo')).toBeTruthy()
    //   expect(importedComponents.has('AddToCart')).toBeTruthy()
    // })

    // it('should collect property definitions', () => {
    //   // TODO: Implement
    //   console.log(propertyDefinitions)
    // })

    // it('should collect injected stores', () => {
    //   // TODO: Implement
    //   console.log(injectedStores)
    // })
  })
})
