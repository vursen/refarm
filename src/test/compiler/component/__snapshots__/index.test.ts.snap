// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store template should parse ast 1`] = `
Array [
  Object {
    "attributes": Array [
      Object {
        "children": Array [],
        "name": undefined,
        "type": "Attribute",
        "value": Object {
          "children": Array [],
          "type": "Text",
          "value": "../components/Photo",
        },
      },
    ],
    "children": Array [],
    "type": "Element",
  },
  Object {
    "children": Array [],
    "type": "Text",
    "value": "
",
  },
  Object {
    "attributes": Array [
      Object {
        "children": Array [],
        "name": undefined,
        "type": "Attribute",
        "value": Object {
          "children": Array [],
          "type": "Text",
          "value": "../components/AddToCart",
        },
      },
    ],
    "children": Array [],
    "type": "Element",
  },
  Object {
    "children": Array [],
    "type": "Text",
    "value": "

",
  },
  Object {
    "attributes": Array [
      Object {
        "children": Array [],
        "name": undefined,
        "type": "Attribute",
        "value": Object {
          "children": Array [],
          "type": "Text",
          "value": "product-page",
        },
      },
    ],
    "children": Array [
      Object {
        "children": Array [],
        "type": "Text",
        "value": "
  ",
      },
      Object {
        "attributes": Array [],
        "children": Array [
          Object {
            "children": Array [],
            "type": "Text",
            "value": "
    ",
          },
          Object {
            "children": Array [],
            "expression": Object {
              "ast": Node {
                "computed": false,
                "end": 137,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 22,
                    "line": 6,
                  },
                  "start": Position {
                    "column": 5,
                    "line": 6,
                  },
                },
                "object": Node {
                  "end": 124,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 9,
                      "line": 6,
                    },
                    "start": Position {
                      "column": 5,
                      "line": 6,
                    },
                  },
                  "start": 120,
                  "type": "ThisExpression",
                },
                "optional": false,
                "property": Node {
                  "end": 137,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 22,
                      "line": 6,
                    },
                    "start": Position {
                      "column": 10,
                      "line": 6,
                    },
                  },
                  "name": "productTitle",
                  "start": 125,
                  "type": "Identifier",
                },
                "start": 120,
                "type": "MemberExpression",
              },
              "children": Array [],
              "type": "Expression",
            },
            "type": "MustacheTag",
          },
          Object {
            "children": Array [],
            "type": "Text",
            "value": "
  ",
          },
        ],
        "type": "Element",
      },
      Object {
        "children": Array [],
        "type": "Text",
        "value": "

  ",
      },
      Object {
        "children": Array [
          Object {
            "attributes": Array [
              Object {
                "children": Array [],
                "name": undefined,
                "type": "Attribute",
                "value": Object {
                  "children": Array [],
                  "expression": Object {
                    "ast": Node {
                      "computed": false,
                      "end": 217,
                      "loc": SourceLocation {
                        "end": Position {
                          "column": 20,
                          "line": 11,
                        },
                        "start": Position {
                          "column": 11,
                          "line": 11,
                        },
                      },
                      "object": Node {
                        "end": 213,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 16,
                            "line": 11,
                          },
                          "start": Position {
                            "column": 11,
                            "line": 11,
                          },
                        },
                        "name": "photo",
                        "start": 208,
                        "type": "Identifier",
                      },
                      "optional": false,
                      "property": Node {
                        "end": 217,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 20,
                            "line": 11,
                          },
                          "start": Position {
                            "column": 17,
                            "line": 11,
                          },
                        },
                        "name": "src",
                        "start": 214,
                        "type": "Identifier",
                      },
                      "start": 208,
                      "type": "MemberExpression",
                    },
                    "children": Array [],
                    "type": "Expression",
                  },
                  "type": "MustacheTag",
                },
              },
              Object {
                "children": Array [],
                "name": undefined,
                "type": "Attribute",
                "value": Object {
                  "children": Array [],
                  "expression": Object {
                    "ast": Node {
                      "computed": false,
                      "end": 239,
                      "loc": SourceLocation {
                        "end": Position {
                          "column": 20,
                          "line": 12,
                        },
                        "start": Position {
                          "column": 11,
                          "line": 12,
                        },
                      },
                      "object": Node {
                        "end": 235,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 16,
                            "line": 12,
                          },
                          "start": Position {
                            "column": 11,
                            "line": 12,
                          },
                        },
                        "name": "photo",
                        "start": 230,
                        "type": "Identifier",
                      },
                      "optional": false,
                      "property": Node {
                        "end": 239,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 20,
                            "line": 12,
                          },
                          "start": Position {
                            "column": 17,
                            "line": 12,
                          },
                        },
                        "name": "alt",
                        "start": 236,
                        "type": "Identifier",
                      },
                      "start": 230,
                      "type": "MemberExpression",
                    },
                    "children": Array [],
                    "type": "Expression",
                  },
                  "type": "MustacheTag",
                },
              },
            ],
            "children": Array [],
            "type": "InlineComponent",
          },
        ],
        "type": "EachBlock",
      },
      Object {
        "children": Array [],
        "type": "Text",
        "value": "

  ",
      },
      Object {
        "attributes": Array [
          Object {
            "children": Array [],
            "name": undefined,
            "type": "Attribute",
            "value": Object {
              "children": Array [],
              "expression": Object {
                "ast": Node {
                  "computed": false,
                  "end": 301,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 29,
                      "line": 17,
                    },
                    "start": Position {
                      "column": 15,
                      "line": 17,
                    },
                  },
                  "object": Node {
                    "end": 291,
                    "loc": SourceLocation {
                      "end": Position {
                        "column": 19,
                        "line": 17,
                      },
                      "start": Position {
                        "column": 15,
                        "line": 17,
                      },
                    },
                    "start": 287,
                    "type": "ThisExpression",
                  },
                  "optional": false,
                  "property": Node {
                    "end": 301,
                    "loc": SourceLocation {
                      "end": Position {
                        "column": 29,
                        "line": 17,
                      },
                      "start": Position {
                        "column": 20,
                        "line": 17,
                      },
                    },
                    "name": "productId",
                    "start": 292,
                    "type": "Identifier",
                  },
                  "start": 287,
                  "type": "MemberExpression",
                },
                "children": Array [],
                "type": "Expression",
              },
              "type": "MustacheTag",
            },
          },
          Object {
            "children": Array [],
            "name": undefined,
            "type": "Attribute",
            "value": Object {
              "children": Array [],
              "expression": Object {
                "ast": Node {
                  "computed": false,
                  "end": 338,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 35,
                      "line": 18,
                    },
                    "start": Position {
                      "column": 18,
                      "line": 18,
                    },
                  },
                  "object": Node {
                    "end": 325,
                    "loc": SourceLocation {
                      "end": Position {
                        "column": 22,
                        "line": 18,
                      },
                      "start": Position {
                        "column": 18,
                        "line": 18,
                      },
                    },
                    "start": 321,
                    "type": "ThisExpression",
                  },
                  "optional": false,
                  "property": Node {
                    "end": 338,
                    "loc": SourceLocation {
                      "end": Position {
                        "column": 35,
                        "line": 18,
                      },
                      "start": Position {
                        "column": 23,
                        "line": 18,
                      },
                    },
                    "name": "productTitle",
                    "start": 326,
                    "type": "Identifier",
                  },
                  "start": 321,
                  "type": "MemberExpression",
                },
                "children": Array [],
                "type": "Expression",
              },
              "type": "MustacheTag",
            },
          },
        ],
        "children": Array [],
        "type": "InlineComponent",
      },
      Object {
        "children": Array [],
        "type": "Text",
        "value": "
",
      },
    ],
    "type": "Element",
  },
]
`;
